### PR TITLE
20607 sync status

### PR DESCRIPTION
**Pull request by ianzhang366**
Addressing [#20607](https://github.com/stolostron/backlog/issues/20607)

With this code in, the hypershiftDeployment status will be like the following:
```
status:
  conditions:
  - lastTransitionTime: "2022-03-28T15:47:53Z"
    message: Retreived secret aws
    observedGeneration: 4
    reason: AsExpected
    status: "True"
    type: ProviderSecretConfigured
  - lastTransitionTime: "2022-03-28T15:48:10Z"
    message: ""
    observedGeneration: 3
    reason: ConfiguredAsExpected
    status: "True"
    type: PlatformInfrastructureConfigured
  - lastTransitionTime: "2022-03-28T15:48:13Z"
    message: ""
    observedGeneration: 4
    reason: ConfiguredAsExpected
    status: "True"
    type: ManifestWorkConfigured
  - lastTransitionTime: "2022-03-28T15:48:13Z"
    message: ""
    observedGeneration: 4
    reason: ConfiguredAsExpected
    status: "True"
    type: PlatformIAMConfigured
  - lastTransitionTime: "2022-03-28T15:48:14Z"
    message: All resources are available
    observedGeneration: 4
    reason: ResourcesAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2022-03-28T15:50:02Z"
    message: Minimum availability requires 2 replicas, current 0 available
    observedGeneration: 4
    reason: WaitingForAvailableMachines
    status: "False"
    type: NodePool
  - lastTransitionTime: "2022-03-28T15:48:14Z"
    message: Apply manifest work complete
    observedGeneration: 4
    reason: AppliedManifestWorkComplete
    status: "True"
    type: Applied
  - lastTransitionTime: "2022-03-28T15:50:01Z"
    message: ""
    observedGeneration: 4
    reason: HostedClusterAsExpected
    status: "True"
    type: HostedCluster
```

There are extra conditions `HostedCluster` and `NodePool`.

The `HostedCluster` will simple copy over the statusFeedback from the manifestwork's hostedcluster status.

The `NodePool` will pop up the first fail nodepool, with status copyed from the manifestwork's nodepool status. If there isn't any failed nodepool, then will use `NodePoolsProvisioned` as reason to tell user that all nodepools are ready.

### CODE REVIEW

**Comment by jnpacker** [comment]
1. What happens if there is more then 1x NodePools?
2. Is it difficult if we brought all status's back from HostedCluster at least to help with problem determination? Creating its own condition? `status.hostedCluster.conditions[ ]`

File: pkg/controllers/manifestwork_test.go [hunk 1]
```diff
@@ -67,7 +78,7 @@ func newManifestResourceChecker(ctx context.Context, clt client.Client, key type
 	for _, w := range wl {
 		u := &unstructured.Unstructured{}
 		if err := json.Unmarshal(w.Raw, u); err != nil {
-			return nil, fmt.Errorf("faield convert manifest to unstructured, err: %w", err)
+			return fmt.Errorf("faield convert manifest to unstructured, err: %w", err)
```
Comment (jnpacker) [inline pkg/controllers/manifestwork_test.go]: typo

**Comment by ianzhang366** [comment]
> What happens if there is more then 1x NodePools?
Is it difficult if we brought all status's back from HostedCluster at least to help with problem determination? Creating its own condition? status.hostedCluster.conditions[ ]

if there is more than one nodepool, it will try to find the failed one, and use it. If there's no failed ones, then it will say all ready.

In terms of bring all status from hostedcluster.status to hypershiftDeployment, we will have to map field by field, would be error-prone tho. I would prefer only to bring key conditions back which would give a user enough info jump to hostedcluster. 

Even if we bring all status from hostedcluster, the user still need to jump to hostedcluster CR for debug tho.

**Comment by ianzhang366** [comment]
New status are:
```
status:
  conditions:
  - lastTransitionTime: "2022-03-28T20:44:28Z"
    message: Retreived secret aws
    observedGeneration: 5
    reason: AsExpected
    status: "True"
    type: ProviderSecretConfigured
  - lastTransitionTime: "2022-03-28T20:55:01Z"
    message: 'Removing AWS infrastructure with infra-id: manifestwork-status-ztmgc'
    observedGeneration: 5
    reason: Destroying
    status: "False"
    type: PlatformInfrastructureConfigured
  - lastTransitionTime: "2022-03-28T20:44:50Z"
    message: ""
    observedGeneration: 4
    reason: ConfiguredAsExpected
    status: "True"
    type: ManifestWorkConfigured
  - lastTransitionTime: "2022-03-28T20:44:50Z"
    message: ""
    observedGeneration: 4
    reason: ConfiguredAsExpected
    status: "True"
    type: PlatformIAMConfigured
  - lastTransitionTime: "2022-03-28T20:44:50Z"
    message: All resources are available
    observedGeneration: 5
    reason: ResourcesAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2022-03-28T20:44:50Z"
    message: Apply manifest work complete
    observedGeneration: 5
    reason: AppliedManifestWorkComplete
    status: "True"
    type: Applied
  - lastTransitionTime: "2022-03-28T20:46:41Z"
    message: ""
    observedGeneration: 5
    reason: HostedClusterAsExpected
    status: "True"
    type: HostedClusterAvaliable
  - lastTransitionTime: "2022-03-28T20:44:50Z"
    message: ""
    observedGeneration: 5
    reason: Partial
    status: "True"
    type: HostedClusterProgress
  - lastTransitionTime: "2022-03-28T20:45:41Z"
    message: Minimum availability requires 2 replicas, current 0 available
    observedGeneration: 5
    reason: WaitingForAvailableMachines
    status: "False"
    type: NodePool
```

**Pull request merged**